### ISSUE TITLE
Add new APIs in support of rsync

### ISSUE THREAD

**Issue by djbrooke** [IQSS/dataverse#3352]
We'll need to add some new APIs to support adding files using rsync.
- Checksum validation API
- Upload Mechanisms API (was Upload API)
- Script Request API

**Comment by pdurbin** [comment]
During the 2016-09-08 SBGrid Sprint Planning meeting ( https://docs.google.com/document/d/1wWSdKUOGA1L7UqFsgF3aOs8_9uyjnVpsPAxk7FObOOI/edit ) this issue was given an effort level of "8". Here are links to the code that was demo'ed the day before:

- Checksum validation API: https://github.com/bmckinney/dataverse-canonical/blob/069fd8d72bfe96ec93b4005550b0c771dee01fc4/src/main/java/edu/harvard/iq/dataverse/api/Datasets.java#L728
- Upload API: ???
- Script Request API: https://github.com/bmckinney/dataverse-canonical/blob/069fd8d72bfe96ec93b4005550b0c771dee01fc4/src/main/java/edu/harvard/iq/dataverse/api/Datasets.java#L828

The biggest unknown for me here is what is meant by the "Upload API". Has it been written already? I'm adding @djbrooke @scolapasta and @bmckinney to help me figure out what the Upload API is.

**Comment by pdurbin** [comment]
@bmckinney @scolapasta @djbrooke and I discussed this and it turns out that what was really meant was Upload Mechanisms API. This is the "setUploadMechanisms" method, for example at https://github.com/bmckinney/dataverse-canonical/blob/069fd8d72bfe96ec93b4005550b0c771dee01fc4/src/main/java/edu/harvard/iq/dataverse/api/Dataverses.java#L775 which is to say that that code has already been written so we should go ahead and make a pull request. I moved this issue to Ready at https://waffle.io/IQSS/dataverse

**Comment by pdurbin** [comment]
@bmckinney and I discussed this issue this morning at https://dataverseteam.slack.com/archives/dcm-merge/p1477917910000400 and I dragged it to "Development" in https://waffle.io/IQSS/dataverse and will assign it to myself since I plan to work on it this week.

**Comment by pdurbin** [comment]
@bmckinney and I discussed this issue today and he said I shouldn't use my old code at https://github.com/pdurbin/dataverse/tree/3145-dcm because he had to tweak it. He's going to leave a comment here about where I should look for the latest code. Thanks!

**Comment by pdurbin** [comment]
Generally speaking, I should be grabbing code from https://github.com/bmckinney/dataverse-canonical/compare/sbgrid-dcm-rsync

For example, here's the `okResponse` returned when the DCM says "validation passed": https://github.com/bmckinney/dataverse-canonical/blob/439c0585cdb0d2af3c68bedb56986f6d8089c665/src/main/java/edu/harvard/iq/dataverse/api/Datasets.java#L951

**Comment by pdurbin** [comment]
I need to switch gears to another issue so I pushed some scratch work code I was working on to https://github.com/pdurbin/dataverse/tree/3352-talk-to-dcm for safekeeping.

One nice thing about spending a little time with this code again is that it motivated me to run the real DCM code at https://github.com/sbgrid/data-capture-module rather than the fake stub I was using to develop against previously. I plan to leave a comment on https://github.com/sbgrid/data-capture-module/issues/4 about having to wait for cron.

@bmckinney said he could pick up this issue after #3352 which sounds fine to me. I'm pulling this issue out of "Development" and putting it back in "Ready" at https://waffle.io/IQSS/dataverse

**Comment by pdurbin** [comment]
I'm picking up work on this again and just pushed some scratch work to https://github.com/pdurbin/dataverse/commit/172b9689d9c077b6cac4d687d877f3f617aad449

Because pull request #3497 has not yet been merged, I had to steal an update to UserNotification.java and I hard-coded "42" rather than calling `jo.start("FileSystemImportJob", props);`. @djbrooke and I agree that the plan is to merge pull request #3497 first. Then I'll merge that code into whatever branch I put through QA for this issue. Next I'm considering making a new branch based on the branch behind pull request #3497 (3353-batch-job-import) so see if that makes more sense, so that I don't have to wait for it to be merged to have a complete solution.

**Comment by djbrooke** [comment]
In the planning meeting today, @pdurbin mentioned that he's blocked by the code in #3353 not currently working. @landreev offered his assistance. 

A helpful concept (to me, at least) is that someone suggested thinking of this as an API wrapper around the rsync functionality.

**Comment by pdurbin** [comment]
I learned in a [meeting yesterday](https://docs.google.com/document/d/1Cp0myGJKAMWQLIT8wHE8JLitUOsySUM0S7yQ_rx4Cgw/edit?usp=sharing) with @pameyer @djbrooke and @scolapasta yesterday that @bmckinney is apparently working on this. It sounds like the plan is to copy and paste working code from https://github.com/sbgrid/sbgrid-dataverse/tree/feature/dcm-api (unless I'm mistaken) to make a pull request against the main IQSS repo.

**Comment by bmckinney** [comment]
The [SBGrid feature overlay](https://github.com/sbgrid/sbgrid-dataverse/tree/feature/dcm-api)  was made into a [Dataverse branch](https://github.com/IQSS/dataverse/tree/3352-dcm-apis) back on 11/4. I just added a couple bug fixes and merged develop.

@pameyer and I have a working DCM at dv.sbgrid.org (basically the demo we showed at the September 2016 meeting). (see: [Testing DCM](https://github.com/sbgrid/sbgrid-dataverse/wiki/HOWTO:-Testing-DCM)) It includes the [UI overlay for DCM](https://github.com/sbgrid/sbgrid-dataverse/tree/feature/dcm-ui) (not yet made into a dataverse branch since it hasn't gone through code review, etc.)

Finally, see [swagger docs](https://dv.sbgrid.org/apidocs/) for more details.

**Comment by bmckinney** [comment]
sorry, didn't mean to close this!

**Comment by pdurbin** [comment]
@bmckinney ah, good to know that the 3352-dcm-apis branch is in play. Thanks for the reminder. I much prefer working from a branch that's in the main IQSS repo rather than the overlay.

**Comment by djbrooke** [comment]
Moving out of development. This is dependent on #3353 and will be started on when that code is merged.

**Comment by djbrooke** [comment]
I'm closing this issue.  We'll create new issues, clarify, and re-estimate if we still need these APIs as part of the new approach for handling large datasets.

**Comment by pdurbin** [comment]
This issue lives on as #3725.
